File: ios/App/App/RustSearchPlugin.swift
import Capacitor
import Foundation
import Darwin

@objc(RustSearchPlugin)
public class RustSearchPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "RustSearchPlugin"
    public let jsName = "RustSearch"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "prepareIndex", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "search", returnType: CAPPluginReturnPromise)
    ]

    private var prepared = false
    private var activeRegion = ""
    private var activeVersion = ""
    private let bridge = RustSearchBridge()

    @objc func prepareIndex(_ call: CAPPluginCall) {
        let regionId = call.getString("regionId")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let graphPath = call.getString("graphPath")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let poiPath = call.getString("poiPath")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let dataVersion = call.getString("dataVersion")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        var response: [String: Any] = [
            "nativeAvailable": bridge.isAvailable
        ]

        guard bridge.isAvailable else {
            prepared = false
            response["prepared"] = false
            response["reason"] = "Rust native library not loaded"
            call.resolve(response)
            return
        }

        let success = bridge.prepareIndex(
            regionId: regionId,
            graphPath: graphPath,
            poiPath: poiPath,
            dataVersion: dataVersion
        )

        prepared = success
        if success {
            activeRegion = regionId
            activeVersion = dataVersion
        }

        response["prepared"] = success
        response["regionId"] = regionId
        response["dataVersion"] = dataVersion
        call.resolve(response)
    }

    @objc func search(_ call: CAPPluginCall) {
        let query = call.getString("query")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let regionId = call.getString("regionId")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? activeRegion
        let limit = max(1, min(call.getInt("limit") ?? 6, 20))
        let biasLng = call.getDouble("biasLng") ?? 0.0
        let biasLat = call.getDouble("biasLat") ?? 0.0

        var response: [String: Any] = [
            "nativeAvailable": bridge.isAvailable,
            "prepared": prepared,
            "results": []
        ]

        guard bridge.isAvailable, prepared, !query.isEmpty else {
            call.resolve(response)
            return
        }

        guard let payload = bridge.search(
            query: query,
            regionId: regionId,
            limit: limit,
            biasLng: biasLng,
            biasLat: biasLat
        ) else {
            call.resolve(response)
            return
        }

        guard
            let data = payload.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            response["error"] = "Rust response could not be parsed"
            call.resolve(response)
            return
        }

        if let results = object["results"] as? [[String: Any]] {
            response["results"] = results
        }
        response["regionId"] = object["regionId"] as? String ?? regionId
        response["dataVersion"] = object["dataVersion"] as? String ?? activeVersion
        response["latencyMs"] = object["latencyMs"] as? Double ?? -1
        call.resolve(response)
    }
}

private final class RustSearchBridge {
    private typealias PrepareIndexFn = @convention(c) (
        UnsafePointer<CChar>?,
        UnsafePointer<CChar>?,
        UnsafePointer<CChar>?,
        UnsafePointer<CChar>?
    ) -> Bool

    private typealias SearchFn = @convention(c) (
        UnsafePointer<CChar>?,
        UnsafePointer<CChar>?,
        Int32,
        Double,
        Double
    ) -> UnsafePointer<CChar>?

    private typealias FreeStringFn = @convention(c) (UnsafeMutablePointer<CChar>?) -> Void

    private let handle: UnsafeMutableRawPointer?
    private let prepareIndexFn: PrepareIndexFn?
    private let searchFn: SearchFn?
    private let freeStringFn: FreeStringFn?

    var isAvailable: Bool {
        handle != nil && prepareIndexFn != nil && searchFn != nil
    }

    init() {
        if let dynamicHandle = RustSearchBridge.openDynamicLibrary() {
            handle = dynamicHandle
        } else if let existing = dlopen(nil, RTLD_NOW) {
            handle = existing
        } else {
            handle = nil
        }

        prepareIndexFn = RustSearchBridge.loadSymbol(
            name: "rust_search_prepare_index",
            handle: handle
        )
        searchFn = RustSearchBridge.loadSymbol(
            name: "rust_search_search",
            handle: handle
        )
        freeStringFn = RustSearchBridge.loadSymbol(
            name: "rust_search_free_string",
            handle: handle
        )
    }

    func prepareIndex(regionId: String, graphPath: String, poiPath: String, dataVersion: String) -> Bool {
        guard let fn = prepareIndexFn else { return false }
        return regionId.withCString { regionCString in
            graphPath.withCString { graphCString in
                poiPath.withCString { poiCString in
                    dataVersion.withCString { versionCString in
                        fn(regionCString, graphCString, poiCString, versionCString)
                    }
                }
            }
        }
    }

    func search(query: String, regionId: String, limit: Int, biasLng: Double, biasLat: Double) -> String? {
        guard let fn = searchFn else { return nil }

        let resultPtr: UnsafePointer<CChar>? = query.withCString { queryCString in
            regionId.withCString { regionCString in
                fn(queryCString, regionCString, Int32(limit), biasLng, biasLat)
            }
        }

        guard let resultPtr else {
            return nil
        }

        let output = String(cString: resultPtr)
        if let freeFn = freeStringFn {
            let mutable = UnsafeMutablePointer(mutating: resultPtr)
            freeFn(mutable)
        }
        return output
    }

    private static func loadSymbol<T>(name: String, handle: UnsafeMutableRawPointer?) -> T? {
        guard let handle else { return nil }
        guard let symbol = dlsym(handle, name) else { return nil }
        return unsafeBitCast(symbol, to: T.self)
    }

    private static func openDynamicLibrary() -> UnsafeMutableRawPointer? {
        let candidates: [String] = [
            "libmelange_rust_search.dylib",
            "melange_rust_search.framework/melange_rust_search"
        ]

        for candidate in candidates {
            if let frameworksPath = Bundle.main.privateFrameworksPath {
                let fullPath = "\(frameworksPath)/\(candidate)"
                if let handle = dlopen(fullPath, RTLD_NOW | RTLD_GLOBAL) {
                    return handle
                }
            }
        }
        return nil
    }
}

File: android/app/src/main/java/com/aimapsystem/app/RustSearchPlugin.java
package com.aimapsystem.app;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONArray;
import org.json.JSONObject;

@CapacitorPlugin(name = "RustSearch")
public class RustSearchPlugin extends Plugin {

    private boolean prepared = false;
    private String activeRegion = "";
    private String activeVersion = "";

    @PluginMethod
    public void prepareIndex(PluginCall call) {
        String regionId = call.getString("regionId", "").trim();
        String graphPath = call.getString("graphPath", "").trim();
        String poiPath = call.getString("poiPath", "").trim();
        String dataVersion = call.getString("dataVersion", "").trim();

        JSObject result = new JSObject();
        result.put("nativeAvailable", RustSearchNativeBridge.isAvailable());

        if (!RustSearchNativeBridge.isAvailable()) {
            prepared = false;
            result.put("prepared", false);
            result.put("reason", "Rust native library not loaded");
            call.resolve(result);
            return;
        }

        try {
            boolean success = RustSearchNativeBridge.nativePrepareIndex(
              regionId,
              graphPath,
              poiPath,
              dataVersion
            );
            prepared = success;
            if (success) {
                activeRegion = regionId;
                activeVersion = dataVersion;
            }
            result.put("prepared", success);
            result.put("regionId", regionId);
            result.put("dataVersion", dataVersion);
            call.resolve(result);
        } catch (Throwable error) {
            prepared = false;
            result.put("prepared", false);
            result.put("error", error.getMessage());
            call.resolve(result);
        }
    }

    @PluginMethod
    public void search(PluginCall call) {
        String query = call.getString("query", "").trim();
        String regionId = call.getString("regionId", activeRegion).trim();
        int limit = call.getInt("limit", 6);
        Double biasLng = call.getDouble("biasLng");
        Double biasLat = call.getDouble("biasLat");

        JSObject result = new JSObject();
        result.put("nativeAvailable", RustSearchNativeBridge.isAvailable());
        result.put("prepared", prepared);
        result.put("results", new JSArray());

        if (!RustSearchNativeBridge.isAvailable() || !prepared || query.isEmpty()) {
            call.resolve(result);
            return;
        }

        try {
            String payload = RustSearchNativeBridge.nativeSearch(
              query,
              regionId,
              Math.max(1, Math.min(20, limit)),
              biasLng == null ? 0.0 : biasLng,
              biasLat == null ? 0.0 : biasLat
            );

            if (payload == null || payload.trim().isEmpty()) {
                call.resolve(result);
                return;
            }

            JSONObject parsed = new JSONObject(payload);
            JSONArray nativeResults = parsed.optJSONArray("results");
            if (nativeResults != null) {
                JSArray out = new JSArray();
                for (int i = 0; i < nativeResults.length(); i++) {
                    out.put(nativeResults.get(i));
                }
                result.put("results", out);
            }

            result.put("regionId", parsed.optString("regionId", regionId));
            result.put("dataVersion", parsed.optString("dataVersion", activeVersion));
            result.put("latencyMs", parsed.optDouble("latencyMs", -1));
            call.resolve(result);
        } catch (Throwable error) {
            result.put("error", error.getMessage());
            call.resolve(result);
        }
    }
}
